### PR TITLE
readOnly writeOnly validation

### PR DESCRIPTION
**Pull request by danicc097**
Adds support for ``readOnly`` and ``writeOnly`` example validation based on request or response context, if any. Will need to fix relevant tests (or perhaps misunderstood something). Again, all comments appreciated

### CODE REVIEW

**Comment by fenollp** [comment]
Ah sorry I missed reviewing this and now this needs rebasing.

File: openapi3/validation_options.go [hunk 1]
```diff
@@ -5,11 +5,23 @@ import "context"
...
+
+// ValidationOptions provides configuration for validating OpenAPI documents.
 type ValidationOptions struct {
 	SchemaFormatValidationEnabled   bool
 	SchemaPatternValidationDisabled bool
-	ExamplesValidationDisabled      bool
+	ExamplesValidation              *ExamplesValidation
+}
+
+func NewValidationOptions() *ValidationOptions {
```
Comment (danicc097) [inline openapi3/validation_options.go:21]: not sure about this

File: openapi3filter/validation_error_test.go [hunk 2]
```diff
@@ -372,15 +372,6 @@ func getValidationTests(t *testing.T) []*validationTest {
 				Title:  `property "name" is missing`,
 				Source: &ValidationErrorSource{Pointer: "/category/tags/0/name"}},
 		},
-		{
```
Comment (danicc097) [inline openapi3filter/validation_error_test.go:375]: can test both read/writeonly in openapi3filter/validate_readonly_writeonly_test.go instead

File: openapi3/response.go [hunk 3]
```diff
@@ -115,6 +115,9 @@ func (response *Response) Validate(ctx context.Context) error {
 	if response.Description == nil {
 		return errors.New("a short description of the response is required")
 	}
+	if xv := getValidationOptions(ctx).ExamplesValidation; !xv.Disabled {
```
Comment (danicc097) [inline openapi3/response.go:118]: not sure if it's a good use of context switching flags back and forth while we traverse the spec

**Comment by danicc097** [comment]
PS: sorry about the useless review code snippets, can't seem to edit the selection after submitting...

File: openapi3filter/validation_error_test.go (hunk 2 above)
Comment (fenollp) [reply openapi3filter/validation_error_test.go:375]: Sounds good to me.

**Comment by danicc097** [comment]
I think this should cover everything: readOnly and writeOnly properties request, response and examples validation

**Comment by danicc097** [comment]
@fenollp does this look good to you, before I fix merge conflicts?

File: openapi3/example_validation.go [hunk 4]
```diff
@@ -1,5 +1,16 @@
 package openapi3
 
-func validateExampleValue(input interface{}, schema *Schema) error {
-	return schema.VisitJSON(input, MultiErrors())
+import "context"
+
+func validateExampleValue(ctx context.Context, input interface{}, schema *Schema) error {
+	opts := make([]SchemaValidationOption, 0, 3)
```
Comment (fenollp) [inline openapi3/example_validation.go:6]:
```suggestion
	opts := make([]SchemaValidationOption, 0, 2)
```

File: openapi3/validation_options.go [hunk 5]
```diff
@@ -3,46 +3,57 @@ package openapi3
...
+type ValidationOption func(options *validationOptions)
 
-// ValidationOptions provide configuration for validating OpenAPI documents.
-type ValidationOptions struct {
+type examplesValidation struct {
+	Disabled     bool
+	AsReq, AsRes bool
+}
+
+type validationOptions struct {
```
Comment (fenollp) [inline openapi3/validation_options.go:13]: Please keep that type exported. It's bad documentation practice to have public types that mention private types.

File: openapi3/validation_options.go [hunk 6]
```diff
@@ -3,46 +3,57 @@ package openapi3
...
 	SchemaFormatValidationEnabled   bool
 	SchemaPatternValidationDisabled bool
-	ExamplesValidationDisabled      bool
+	ExamplesValidation              *examplesValidation
```
Comment (fenollp) [inline openapi3/validation_options.go:16]: I'm working on a PR that privatized these fields. Please privatize yours.

File: openapi3/validation_options.go [hunk 7]
```diff
@@ -3,46 +3,57 @@ package openapi3
...
+}
+
+// NewValidationOptions creates configuration for validating OpenAPI documents.
+func NewValidationOptions() *validationOptions {
```
Comment (fenollp) [inline openapi3/validation_options.go:20]: Remove this

**Review by fenollp — CHANGES_REQUESTED** [review]

File: openapi3/schema.go [hunk 8]
```diff
@@ -1456,19 +1458,28 @@ func (schema *Schema) visitJSONObject(settings *schemaValidationSettings, value
...
 			}
+
+			if value[propName] != nil {
+				if reqRO {
+					me = append(me, fmt.Errorf("readOnly property %q in request", propName))
+				} else if repWO {
+					me = append(me, fmt.Errorf("writeOnly property %q in response", propName))
+				}
+			}
 		}
```
Comment (danicc097) [inline openapi3/schema.go:1480]: This is now fixed and tested, I messed up earlier. I don't think anything apart from default values will break ro/wo validation?

File: openapi3/validation_options.go [hunk 9]
```diff
@@ -5,33 +5,34 @@ import "context"
...
 
-// ValidationOptions provide configuration for validating OpenAPI documents.
+// ValidationOptions provides configuration for validating OpenAPI documents.
 type ValidationOptions struct {
-	SchemaFormatValidationEnabled   bool
-	SchemaPatternValidationDisabled bool
-	ExamplesValidationDisabled      bool
+	schemaFormatValidationEnabled                    bool
+	schemaPatternValidationDisabled                  bool
+	examplesValidationDisabled                       bool
```
Comment (fenollp) [inline openapi3/validation_options.go:12]: Please stop touching things not related to your PR. I was asking that the fields you introduce be private, the other ones will come later in another PR.

File: openapi3/schema.go (hunk 8 above)
Comment (fenollp) [reply openapi3/schema.go:1480]: What do you mean?

**Review by fenollp — CHANGES_REQUESTED** [review]
Almost there!

File: openapi3/schema.go (hunk 8 above)
Comment (danicc097) [reply openapi3/schema.go:1480]: Defaults were set originally without taking the property fields into account and I didn't address that. Now if e.g. we are making a request, a default for a readOnly won't be set

**Pull request merged**